### PR TITLE
[Core] Add option to override environment variables for tasks and actors

### PR DESCRIPTION
**Pull request by architkulkarni**
This PR adds the option `override_environment_variables=dict()` to override environment variables for tasks and actors.  These overrides are inherited by nested tasks and nested actors.

This PR builds off of https://github.com/ray-project/ray/pull/11040, and the context is the same:

> ### Context
> For Serve we want to solve the issue of environment isolation. A user should be able to submit two models from to different environments to Serve. For example, one model depends on Tensorflow 1 and the other depends on Tensorflow 2. 
> 
> Our original thinking was to use multi-tenancy and job config to capture the driver environment and let the worker to have the same environment.
> 
> 1. User activate tensorflow 1 environment, runs deploy script in it, the model actor should run in tf1 environment.
> 2. User activate tensorflow 2 environment, ..., the model actor should run in tf2 environment.
> 
> 
> ### Issue with current multi-tenancy
> In Serve, [our architecture](https://docs.ray.io/en/master/serve/architecture.html) involves a centralized controller actor that manages actor creation and destruction. This is a detached actor is started by a different driver. Therefore, when this actor calls `ModelWorkerClass.remote()`. The model worker inherits the job environment of the controller. 
> 
> ```
> Job #1 ---> starts controller
> Job #2 -> driver deploy tf1 -> send to controller -> create tf1 worker (but has the env of job #1!)
> Job #3 -> driver deploy tf2 -> send to controller -> create tf2 worker (but has the env of job #1!)
> ```
> 
> ### Current fix
> Current fix is simply to allow controller override the environment variable for the model workers. The workers created by the controller will still belong to the same job as the controller.
> 
> If you have any suggestion for better way to doing this, please let us know!

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ericl** [comment]
@edoakes did we decide against more general env_id support?

**Comment by edoakes** [comment]
@ericl yeah we decided to go with the most minimal solution for now and @raulchen et. al preferred putting this info in the task spec for simplicity.

I think this makes the most sense for this change in isolation and it should be straightforward to make it more general / pull it out into the GCS in the future. It could make sense to at least abstract it out into an "environment" proto in this PR, not sure.

**Review by ericl — APPROVED** [review]

**Comment by ericl** [comment]
Sounds good

File: src/ray/core_worker/common.h [hunk 1]
```diff
@@ -122,6 +133,9 @@ struct ActorCreationOptions {
   /// When true, the child task will always scheduled on the same placement group
   /// specified in the PlacementOptions.
   bool placement_group_capture_child_tasks = true;
+  /// Environment variables to update for this actor.  Maps a variable name to its
+  /// value.  Can override existing environment variables and introduce new ones.
```
Comment (edoakes) [inline src/ray/core_worker/common.h:137]: Consider adding a comment that these will be propagated to any child actors and/or tasks.

**Review by edoakes — APPROVED** [review]
LGTM. Can you please verify that it shows up in the API ref in the documentation before we merge?

**Pull request merged**